pkg/provider/device42/device42_api.go

```diff
@@ -0,0 +1,145 @@
+	password string
+	username string
+	context  context.Context
+}
+
+type D42PasswordResponse struct {
+	Passwords []D42Password
+}
+
+type D42Password struct {
+	Username     string `json:"username"`
+	LastPwChange string `json:"last_pw_change"`
+	Notes        string `json:"notes"`
+	Label        string `json:"label"`
+	FirstAdded   string `json:"first_added"`
+	Password     string `json:"password"`
+	ID           int    `json:"id"`
+}
+
+func NewAPI(ctx context.Context, baseURL, username, password, hostPort string) *API {
+	api := &API{
+		baseURL:  baseURL,
+		hostPort: hostPort,
+		username: username,
+		password: password,
+		context:  ctx,
+	}
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{MinVersion: tls.VersionTLS12},
+	}
+
+	api.client = &http.Client{Transport: tr}
+	return api
+}
+
+func (api *API) doAuthenticatedRequest(r *http.Request) (*http.Response, error) {
+	r.SetBasicAuth(api.username, api.password)
+	r = r.WithContext(api.context)
+	return api.client.Do(r)
+}
+
+func ReadAndUnmarshal(resp *http.Response, target any) error {
+	var buf bytes.Buffer
+	defer func() {
+		if resp.Body != nil {
+			err := resp.Body.Close()
+			if err != nil {
+				return
+			}
+		}
+	}()
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		return fmt.Errorf("failed to authenticate with the given credentials: %d %s", resp.StatusCode, buf.String())
+	}
+	_, err := buf.ReadFrom(resp.Body)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(buf.Bytes(), target)
+}
+
+func (api *API) GetSecret(secretID string) (D42Password, error) {
+	// https://api.device42.com/#!/Passwords/getPassword
+	endpointURL := fmt.Sprintf("https://%s:%s/api/1.0/passwords/?id=%s&plain_text=yes", api.baseURL, api.hostPort, secretID)
+
+	readSecretRequest, err := http.NewRequest("GET", endpointURL, http.NoBody)
+	if err != nil {
+		return D42Password{}, fmt.Errorf("error: creating secrets request: %w", err)
+	}
+
+	respSecretRead, err := api.doAuthenticatedRequest(readSecretRequest) //nolint:bodyclose // linters bug
+	if err != nil {
+		return D42Password{}, fmt.Errorf(DoRequestError, err)
+	}
+
+	d42PasswordResponse := D42PasswordResponse{}
+	err = ReadAndUnmarshal(respSecretRead, &d42PasswordResponse)
+	if err != nil {
+		return D42Password{}, fmt.Errorf(errJSONSecretUnmarshal, err)
+	}
+	if len(d42PasswordResponse.Passwords) == 0 {
+		return D42Password{}, err
+	}
+	// There should only be one response
+	return d42PasswordResponse.Passwords[0], err
+}
+
+func (api *API) GetSecretMap(_ context.Context, _ esv1beta1.ExternalSecretDataRemoteRef) (map[string][]byte, error) {
+	return nil, fmt.Errorf(errNotImplemented)
+}
+
+func (s D42Password) ToMap() map[string][]byte {
+	m := make(map[string][]byte)
+
+	m["username"] = []byte(s.Username)
+	m["last_pw_change"] = []byte(s.LastPwChange)
+	m["notes"] = []byte(s.Notes)
+	m["label"] = []byte(s.Label)
+	m["first_added"] = []byte(s.FirstAdded)
+	m["password"] = []byte(s.Password)
+	m["id"] = []byte(strconv.Itoa(s.ID))
```

Comment:
These should either be configurable, or be in a `const` somewhere.

Reply:
We really only use the `id` and `password` - I just removed all the other properties in the struct since we don't use them, which should make things cleaner and less expensive to Unmarshal